Mage.Verify/src/test/java/mage/verify/VerifyCardDataTest.java

```diff
@@ -1644,19 +1649,102 @@ private void checkCardCanBeCopied(Card card1) {
             fail(card1, "copy", "throws on copy : " + err.getClass() + " : " + err.getMessage() + " : ");
             return;
         }
+        
+        compareClassRecursive(card1, card2, card1, "[Card", 10);
+    }
 
-        // Checks that ability and effect are of the same class when copied.
-        for (int i = 0; i < card1.getAbilities().size(); i++) {
-            Ability ability1 = card1.getAbilities().get(i);
-            Ability ability2 = card2.getAbilities().get(i);
-            if (!ability1.getClass().equals(ability2.getClass())) {
-                fail(card1, "copy", " miss copy in ability " + ability1.getClass().getName());
+    private void compareClassRecursive(Object obj1, Object obj2, Card originalCard, String msg, int maxDepth) {
+        if (obj1 == null && obj2 == null) {
+            return;
+        } else if (obj1 == null || obj2 == null) {
+            fail(originalCard, "copy", "not same class for " + msg + "]");
+        } else if (obj1.getClass() != obj2.getClass()) {
+            fail(originalCard, "copy", "not same class for " + msg + "<" + obj1.getClass() + ">" + "]");
+        } else if (obj1 == obj2) { // for instances mostly
+            return;
+        } else {
+            // Only recurse so much.
+            if (maxDepth == 0) {
+                return;
             }
-            for (int j = 0; j < ability1.getEffects().size(); j++) {
-                Effect effect1 = ability1.getEffects().get(j);
-                Effect effect2 = ability2.getEffects().get(j);
-                if (!effect1.getClass().equals(effect2.getClass())) {
-                    fail(card1, "copy", "miss copy in effect " + effect1.getClass().getName());
+            // Only recurse on those objects
+            if (obj1 instanceof MageObject || obj1 instanceof Filter || obj1 instanceof Condition || obj1 instanceof Effect
+                    || obj1 instanceof Ability || obj1 instanceof Mana || obj1 instanceof Cost || obj1 instanceof DynamicValue
+                    || obj1 instanceof Choice || obj1 instanceof TargetPointer) {
+                //System.out.println(msg);
```

Comment:
Could check for a non-public constructor that takes a final parameter of the same class right there.